pkg/build/generator/generator.go

```diff
@@ -0,0 +1,262 @@
+package generator
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/golang/glog"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+// BuildGenerator is a central place responsible for generating new Build objects
+// from BuildConfigs and other Builds.
+type BuildGenerator struct {
+	Client GeneratorClient
+}
+
+type GeneratorClient interface {
+	GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuild(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+type Client struct {
+	GetBuildConfigFunc     func(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfigFunc  func(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuildFunc           func(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuildFunc        func(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepositoryFunc func(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+func (c Client) GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error) {
+	return c.GetBuildConfigFunc(ctx, name)
+}
+
+func (c Client) UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error {
+	return c.UpdateBuildConfigFunc(ctx, buildConfig)
+}
+
+func (c Client) GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error) {
+	return c.GetBuildFunc(ctx, name)
+}
+
+func (c Client) CreateBuild(ctx kapi.Context, build *buildapi.Build) error {
+	return c.CreateBuildFunc(ctx, build)
+}
+
+func (c Client) GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error) {
+	return c.GetImageRepositoryFunc(ctx, name)
+}
+
+// Instantiate returns new Build object based on a BuildRequest object
+func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRequest) (*buildapi.Build, error) {
+	glog.V(4).Infof("Generating build from config %s", request.Name)
+	bc, err := g.Client.GetBuildConfig(ctx, request.Name)
+	if err != nil {
+		return nil, err
+	}
+	newBuild, err := g.generateBuildFromConfig(ctx, bc, request.Revision)
+	if err != nil {
+		return nil, err
+	}
+	if err := g.Client.UpdateBuildConfig(ctx, bc); err != nil {
```

Comment:
rather than having to call update here, can you not update the buildconfig in the getNextBuildName method directly instead?  it worries me that you're relying on code that calls getNextBuildName to perform an update afterwards...i thought we agreed that was bad.